byterun/domain.c

```diff
@@ -324,16 +336,31 @@ CAMLprim value caml_domain_spawn(value callback)
   if (p.newdom) {
     /* successfully created a domain.
        p.callback is now owned by that domain */
-    pthread_detach(th);
   } else {
     /* failed */
     void* r;
-    pthread_join(th, &r);
+    pthread_join(*th, &r);
     caml_delete_root(p.callback);
     caml_failwith("failed to allocate domain");
   }
 
-  CAMLreturn (Val_unit);
+  CAMLreturn (th_val);
+}
+
+CAMLprim value caml_ml_domain_join(value domain)
+{
+    CAMLparam1 (domain);
+    pthread_t *th = (pthread_t*) Data_custom_val(domain);
+    int err;
+
+    if (!th)
+        caml_invalid_argument("No thread");
+    caml_gc_log("Domain joining");
+    err = pthread_join(*th, NULL);
```

Comment:
This is a slow operation, so it needs to be in a blocking section (surrounded by `caml_enter_blocking_section` and `caml_leave_blocking_section`). Because `*th` can move, the dereference must be outside the blocking section. (The easiest way to do this is to write `pthread_t th = *(...)` instead of `pthread_t *th = (...)`)